### PR TITLE
Promise support for Nano - Issue 98

### PR DESCRIPTION
**Pull request by glynnbird**
## Overview

Aimed at a v7.x release of Nano, this PR makes Nano return a _Promise_ instead of a _Request_ object as its return value. See https://github.com/apache/couchdb-nano/issues/98. This brings Nano into the modern world without breaking code that uses the callback style. It only breaks code that relied on the events emitted from the Request object or that piped the stream to another stream.

Work done in this PR:

- most functions return a _Promise_
- new ...AsStream| functions are availble for operations that return lots of data. These return a _Request_ object
- test suite updated
- README changed to reflect the new API
- TypeScript definitions updated too
- written `migration_6_to_7` guide

## Testing recommendations

The tests have been updated to reflect the change. Instantiate a Nano instance and most operations will return a _Promise_:

```js
var nano = require('.')
var n = nano('http://localhost:5984')
var db = n.db.use('animals')
db.list({start_key:'cat1',end_key:'cat4'})
  .then(console.log)
  .catch(console.error)
```

## GitHub issue number

https://github.com/apache/couchdb-nano/issues/98

## Related Pull Requests

https://github.com/apache/couchdb-nano/pull/88

## Checklist

- [x] Code is written and works correctly;
- [x] Changes are covered by tests;
- [x] Documentation reflects the changes;

### CODE REVIEW

File: lib/nano.js [hunk 1]
```diff
@@ -68,6 +67,89 @@ module.exports = exports = nano = function dbScope(cfg) {
     }
     return str;
   }
+  var responseHandler = function(req, opts, resolve, reject, callback) {
+    
+    return function(e, h, b) {
+      var parsed;
```
[the remark this answers is not in the document]
Comment (glynnbird) [reply lib/nano.js]: I wasn't sure how far to go with the modernisation of the code. If we're going to use `const` here, then I'll take a look at doing `const`/`let` everywhere else too.

[the remark this answers is not in the document]
Comment (garrensmith) [reply lib/nano.js]: Well maybe do that in a separate PR.
Lets get this landed. Then do a modernisation and then do a release for 7.0.0

File: lib/nano.js [hunk 2]
```diff
@@ -68,6 +67,89 @@ module.exports = exports = nano = function dbScope(cfg) {
...
+      var rh = h && h.headers || {};
```
[the remark this answers is not in the document]
Comment (glynnbird) [reply lib/nano.js]: 👍 This e/h/b business is historic. I've refactored this function to aid readability.

File: lib/nano.js [hunk 3]
```diff
@@ -68,6 +67,89 @@ module.exports = exports = nano = function dbScope(cfg) {
...
+          callback(ret_e);
+        }
+        return ;
+      }
+
+      delete rh.server;
+      delete rh['content-length'];
+
+      if (opts.dontParse) {
+        parsed = b;
```
[the remark this answers is not in the document]
Comment (glynnbird) [reply lib/nano.js]: 👍

File: lib/nano.js [hunk 4]
```diff
@@ -190,67 +269,13 @@ module.exports = exports = nano = function dbScope(cfg) {
 
     log(req);
 
-    if (!callback) {
-      return httpAgent(req);
+    if (opts.stream) {
+      return httpAgent(req, responseHandler(req, opts, null, null, callback));
```
[the remark this answers is not in the document]
Comment (glynnbird) [reply lib/nano.js]: 👍

File: lib/nano.js [hunk 5]
```diff
@@ -471,12 +514,27 @@ module.exports = exports = nano = function dbScope(cfg) {
...
+      return new Promise(function(resolve, reject) {
+        relax({
+          db: dbName,
+          doc: docName,
+          method: 'HEAD',
+          qs: {}
+        }, function(err, body, headers) {
+          if (err) {
+            if (callback) {
+              callback(err);
```
[the remark this answers is not in the document]
Comment (glynnbird) [reply lib/nano.js]: Good point 👍

File: lib/nano.js [hunk 6]
```diff
@@ -612,6 +688,12 @@ module.exports = exports = nano = function dbScope(cfg) {
...
     }
 
+    // http://docs.couchdb.org/en/latest/api/ddoc/views.html#post--db-_design-ddoc-_view-view
+    function viewDocsAsStream(ddoc, viewName, qs, callback) {
+      return view(ddoc, viewName, {type: 'view', stream: true}, qs, callback);
+    }
+    
+
     // geocouch
     function viewSpatial(ddoc, viewName, qs, callback) {
```
[the remark this answers is not in the document]
Comment (glynnbird) [reply lib/nano.js]: I'll create an issue to track this.https://github.com/apache/couchdb-nano/issues/103

File: lib/nano.js [hunk 7]
```diff
@@ -738,6 +825,25 @@ module.exports = exports = nano = function dbScope(cfg) {
       }, callback);
     }
 
+    function insertAttAsStream(docName, attName, att, contentType, qs, callback) {
+      if (typeof qs === 'function') {
```
[the remark this answers is not in the document]
Comment (glynnbird) [reply lib/nano.js]: 👍 done

File: migration_6_to_7.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+var db = nano.db.use('mydb')
+var x = db.get('mydoc') 
+// x is a request object
+```
+
+In Nano 7:
+
+```js
+// Nano 7
+var db = nano.db.use('mydb')
```
[the remark this answers is not in the document]
Comment (glynnbird) [reply migration_6_to_7.md]: 👍

File: migration_6_to_7.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+var x = db.get('mydoc')
```
[the remark this answers is not in the document]
Comment (glynnbird) [reply migration_6_to_7.md]: 👍

File: .jshintrc [hunk 10]
```diff
@@ -3,11 +3,12 @@
   "node":      true,
   "strict":    true,
   "smarttabs": true,
-  "maxlen":    80,
+  "maxlen":    100,
```
Comment (janl) [inline .jshintrc]: naughty!

Comment (glynnbird) [reply .jshintrc]: 😊

File: README.md [hunk 11]
```diff
@@ -109,87 +139,80 @@ In `nano` the callback function receives always three arguments:
...
+}).then((response) =>  {
+  alice = nano.use('alice')
+  return alice.insert({ happy: true }, 'rabbit')
+}).then((response) => {
+  console.log('you have inserted a document with an _id of rabbit')
+  console.log(response);
+})
+```
+
+or in the `async/await` style:
```
Comment (janl) [inline README.md]: nickpick, but should we make this the first example and show raw promises second? The code is just so much nicer :)

**Comment by janl** [comment]
@glynnbird how does this relate to https://github.com/apache/couchdb-nano/pull/88?

**Comment by glynnbird** [comment]
@janl The idea of #88 was to make it so that Promises could be switched on with a config flag. A neat idea, but it plays havoc when you're using it with TypeScript which likes functions to return a fixed type, so instead I wrote-up issue #98 to make Promises the default - to mirror the expectations of the modern JS/TS user. That way the return values would always be Promises, which keeps TS happy and added some extra functions which return streams to the streamy stuff. So, in short, #88 was the inspiration, but it slightly different approach won the day.

**Comment by janl** [comment]
@glynnbird thanks for the context!

File: lib/nano.js [hunk 12]
```diff
@@ -68,6 +78,90 @@ module.exports = exports = nano = function dbScope(cfg) {
     }
     return str;
   }
+  const responseHandler = function(req, opts, resolve, reject, callback) {
+    
+    return function(err, response = { statusCode: 500 }, body = '') {
+      let parsed;
+      const responseHeaders = Object.assign({
```
Comment (garrensmith) [inline lib/nano.js]: Thanks this is much clearer to read.

**Pull request merged**